Bytesections: cleanup API

hhugo opened the pull request:
- Writting and reading `bytesections` no longer interfere (global states are removed) 
- Reading a `bytesections` now returns more information (a `section_table`). In particular, users now have access to the starting position of the sections and no longer have to compute it by accumulating individual section lengths.  
```ocaml
type section_entry = {
  name : string;
  pos  : int;
  len  : int;
}

type section_table = section_entry list
```
- Introduce a new module `Bytesections.Name` that expose all the section names used in the codebase. It reduces the risk of typos and gives a natural location to document all sections. 
- remove `toc` (`read_toc` now return the `section_table`)
- ~~remove `pos_first_section` which was mainly used to recompute section positions.~~
- remove `reset` as we no longer use global state
- fix the codebase accordingly
- ~~fix a bug in tools/stripdebug.ml (the bug doesn't cause problem because of the order in which the sections are encountered)~~

Commit:
Bytesections: cleanup API

hhugo commented on tools/stripdebug.ml:
```diff
@@ -23,27 +23,24 @@ open Misc
   List.iter
-    (fun (name, len) ->
-      if name = "DBUG" || name = "CRCS" then begin
-        seek_in ic (in_channel_length ic + len)
```
This looks like a bug.

dra27 replied:
It is, yes - I'd hit while working on the relocatable patches. In practice it doesn't cause problem because of the order in which the sections are encountered. I'd just switched to `pos_in`, which it looks like is what was intended.

hhugo replied:
The new implementation always seek to the beginning of the section before copying the data.

hhugo replied:
This was fixed by @xavierleroy in f2c2c15dd1123436073398c002e74bf2d1044109

hhugo commented:
I've added checks for invariants and made some type private.

Also introduced a new `Bytesections.Name` submodule which gives us a natural place for documenting various sections.

hhugo commented:
@lthls, @gasche your names are already in the change log entry :). Maybe you want to take a look at this ?

hhugo commented:
I've rebased this PR and improved the description.

nojb commented on bytecomp/bytesections.ml:
```diff
@@ -53,49 +87,50 @@ let read_toc ic =
   if header <> Config.exec_magic_number then raise Bad_magic_number;
-  seek_in ic (pos_trailer - 8 * num_sections);
-  section_table := [];
+  let toc_pos = (pos_trailer - 8 * num_sections) in
```
```suggestion
  let toc_pos = pos_trailer - 8 * num_sections in
```

nojb commented on bytecomp/bytesections.ml:
```diff
@@ -15,31 +15,65 @@
+  let dbug = of_string "DBUG"
+  let crcs = of_string "CRCS"
+  let rntm = of_string "RNTM"
+end
```
My first reaction would have been to use a sum type
```
type name = CODE | DLPT | DLLS | ...
```
which could be exposed directly, together with string conversions with would be used only internally. Also, this makes it nicer to pattern match on the name.

hhugo replied:
I though about it but the current API allows one to store/read arbitrary section which I though could be useful.
One could use `type name = CODE | DLPT | DLLS | ... | Other of string` but then parttern marching is no longer nice because you might want to match `| CODE | Other "CODE"`. Also, if you don't want to allow `Other of string`,  you'll have to fail if you read an unrecognize section name.

nojb replied:
I don't understand why you would need to pattern match on `CODE | Other "CODE"`: the `of_string` function will enforce that it will never return `Other "CODE"`, right? As far as I can see, having a `Other` type would be fine and permit using a sum type both for constructing and pattern matching on section names.

I don't want to insist too much on this, but I find the current code that uses module values for the standard section names but that still pattern matches on the naked string literals a bit ugly.

hhugo replied:
I've pushed a commit to use variants

nojb commented on bytecomp/bytesections.ml:
```diff
@@ -53,49 +87,50 @@ let read_toc ic =
-  seek_sec (in_channel_length ic - 16 - 8 * List.length !section_table)
-           !section_table
+let seek_section t ic name =
+  assert (String.length name = 4);
```
I think this assert is no longer necessary now that you use an abstract type (and if you switch to a sum type, it won't be applicable either).

hhugo replied:
removed

nojb reviewed (commented):
Thanks for the PR, I think it is a nice improvement and the change seems correct. I also left a suggestion: using a sum type for `Name.t` instead of a (private) string.

shindere commented:
Nicolás Ojeda Bär (2023/03/06 09:34 -0800):
> @nojb commented on this pull request.
> 
> 
> 
> > +module Name = struct
> +  type t = string
> +  let of_string name =
> +    if String.length name <> 4 then
> +      invalid_arg "Bytesections.Name.of_string: must be of size 4";
> +    name
> +  let code = of_string "CODE"
> +  let dlpt = of_string "DLPT"
> +  let dlls = of_string "DLLS"
> +  let data = of_string "DATA"
> +  let prim = of_string "PRIM"
> +  let symb = of_string "SYMB"
> +  let dbug = of_string "DBUG"
> +  let crcs = of_string "CRCS"
> +  let rntm = of_string "RNTM"
> +end
> 
> I don't understand why you would need to pattern match on `CODE | Other "CODE"`: the `of_string` function will enforce that it will never return `Other "CODE"`, right? As far as I can see, having a `Other` type would be fine and permit using a sum type both for constructing and pattern matching on section names.
> 
> I don't want to insist too much on this, but I find the current code
> that uses module values for the standard section names but that still
> pattern matches on the naked string literals a bit ugly.

It's true that we already have the backend_type in Syswhich I think is
similar.

nojb commented on bytecomp/bytesections.ml:
```diff
@@ -15,31 +15,94 @@
-  section_table := [];
+    (fun {name; pos = _; len} ->
+       let name = Name.to_string name in
+       assert (String.length name = 4);
```
```suggestion
```

gasche replied:
I would be a bit careful here. I think that the `assert` is not meant to prevent a failure condition or explain a size manipulation in this function, but rather to justify the constant `4` that appears in the matching `read_toc` function below. You can read both in parallel and check that one inverts the other. Now we could avoid the `4` appearing on the other end (notably by using `Name.read ic` instead of `Name.of_string (really_input_string ic 4)`, but as long as one is there we probably want to keep the other.

nojb replied:
Good point, let's disregard this suggestion.

nojb commented on bytecomp/bytelink.ml:
```diff
@@ -371,37 +371,37 @@ let link_bytecode ?final_name tolink exec_name standalone =
        output_byte outchan Opcodes.opSTOP;
        output_byte outchan 0; output_byte outchan 0; output_byte outchan 0;
-       Bytesections.record outchan "CODE";
+       Bytesections.record toc_writer Bytesections.Name.CODE;
```
Nit: in fact I think you can write just `CODE` here and in every other callsite, the compiler will disambiguate the constructor.

hhugo replied:
applied

nojb commented on debugger/symbols.ml:
```diff
@@ -56,16 +56,19 @@ let relocate_event orig ev =
-    raise Toplevel
-  end;
+  let toc =
+    begin try
```
```suggestion
    try
```

hhugo replied:
applied

nojb commented on debugger/symbols.ml:
```diff
@@ -56,16 +56,19 @@ let relocate_event orig ev =
+    with Bytesections.Bad_magic_number | Not_found ->
+      prerr_string bytecode_file; prerr_endline " is not a bytecode file.";
+      raise Toplevel
+    end
```
```suggestion
```

hhugo replied:
applied

nojb reviewed (approved):
LGTM

hhugo commented:
I've rebased and squashed commits.

hhugo commented:
Are more reviews/approvals needed for this one ?

nojb commented:
Merged, thanks!
